### ISSUE TITLE
Possible permanent disconnect with external authenticators

### ISSUE THREAD

**Issue by rafzi** [crossbario/crossbar#560]
When using external authenticators (in other crossbar instances) with the router, there are time windows when connection attempts will be failed with ``no callee registered for procedure 'xxx.internal.authenticate.user.wampcra'``, because the component providing the authenticator has not fully started yet.

The problem is that this kind of error is regarded as fatal by clients and they stop reconnecting. As a result unattended peers could be permanently disconnected when for example restarting the router.

As a workaround we have to react to wamp.error.no_such_procedure as a retryable error instead of fatal which does not feel correct. This workaround only applies if you control client code.

AutobahnJS also completly stops the reconnect attempts when getting no_such_procedure in this scenario.

The router should either refuse connections if it does not have the required authenticators or return a new error class.

**Comment by meejah** [comment]
Perhaps startup behavior should be (slightly) modified to await the "ready" event from component sessions as they start up? That is, when starting from configuration, things like ``start_router_component`` shouldn't return until the ``ready`` event issues from the component. This only happens after ``onJoin`` has run, so an authenticator that ``.register()``s in onJoin would successfully register its method before the next component starts up etc. /cc @oberstet

**Comment by oberstet** [comment]
@meejah 

> things like start_router_component shouldn't return until the ready event issues from the component

This makes sense. With CB standalone, the node configuration should be booted sequentially anyway, and for CB managed, CDC can still choose to run stuff in parallel (via multiple concurrent calls into `start_router_component`. Container worker component starting should work similar.

The question is: what about guest workers? How would a guest worker signal "readiness" in a simple way?

**Comment by meejah** [comment]
Yes, guest workers are trickier (for the record, we currently publish "on_component_started" when the connection is established). I will look into what we could do here.

**Comment by meejah** [comment]
Without requiring guest-worker authors to "do something", I don't think anything can be done.

However, the "native" workers (containers, router-local components) already publish ``crossbar.node.{node_id}.on_worker_ready`` -- so *perhaps* we could tie in with that. The issue here would be: how to know that the guest will be doing this (i.e. should we wait for such an event, or just declare we're ready on connect); how to pass along the information to the guest (e.g. env-var like "CB_READY_URI"); adding configuration options.

**Comment by meejah** [comment]
Maybe an easier solution would be: add a configuration option for guests that is the expected URI to which they will publish when the worker is prepared (i.e. the user makes this up). If there is no such option present, we do as now and declare the guest ready when the WAMP session connects. If such an option *is* present, then we await an event published to that URI before declaring that guest-worker "started".

**Comment by meejah** [comment]
Theoretically, we should be able to tie into the "router side" of any WAMP connections the guest makes, but that has its own issues: the guest might not make any WAMP connection at all, or might make several (and at potentially arbitrary times).

From the router side, we should be able to tie a WAMP connection to a guest by introspecting the PID of the process that initiated the connection (at least, this should be possible on UNIX-y systems, not sure about windows) and comparing to our guest processes.
